saleor/discount/migrations/tasks/saleor3_16.py

```diff
@@ -278,28 +292,54 @@ def migrate_translations(sale_ids, saleid_promotion_map):
 
 
 def migrate_checkout_line_discounts(sale_ids, rule_by_channel_and_sale):
-    if checkout_line_discounts := CheckoutLineDiscount.objects.filter(
+    checkout_line_discounts = CheckoutLineDiscount.objects.filter(
         sale_id__in=sale_ids
-    ).select_related("line__checkout"):
-        for checkout_line_discount in checkout_line_discounts:
+    ).exclude(type="promotion")[:CHECKOUT_LINE_DISCOUNT_BATCH_SIZE]
+    discount_line_ids = list(checkout_line_discounts.values_list("id", flat=True))
+    if discount_line_ids:
+        line_discounts = CheckoutLineDiscount.objects.filter(
+            id__in=discount_line_ids
+        ).select_related("line__checkout")
+        discounts_to_update = []
+        for checkout_line_discount in line_discounts:
+            discounts_to_update.append(checkout_line_discount)
+            checkout_line_discount.type = "promotion"
             if checkout_line := checkout_line_discount.line:
                 channel_id = checkout_line.checkout.channel_id
                 sale_id = checkout_line_discount.sale_id
                 lookup = f"{channel_id}_{sale_id}"
-                checkout_line_discount.type = "promotion"
                 if promotion_rule := rule_by_channel_and_sale.get(lookup):
                     checkout_line_discount.promotion_rule = promotion_rule
 
         CheckoutLineDiscount.objects.bulk_update(
-            checkout_line_discounts, ["promotion_rule_id", "type"]
+            discounts_to_update, ["promotion_rule_id", "type"]
         )
+        migrate_checkout_line_discounts(sale_ids, rule_by_channel_and_sale)
 
 
-def migrate_order_line_discounts(sale_ids, rule_by_channel_and_sale):
+def migrate_order_line_discounts(
+    sale_ids, rule_by_channel_and_sale, last_created_at=None, already_processed_ids=None
+):
     global_pks = [graphene.Node.to_global_id("Sale", pk) for pk in sale_ids]
-    if order_lines := OrderLine.objects.filter(sale_id__in=global_pks).prefetch_related(
-        "order"
-    ):
+    filter_lookup = {"sale_id__in": global_pks}
+    # We need to ensure that all instances with the same created_at timestamp
+    # are processed. We need to prevent a situation when the batch proceeds only
+    # part of the instanced with the same created_at value.
+    # To prevent such situation the `gte` is used in filters with excluding
+    # of already processed ids for this created_at timestamp.
+    if last_created_at:
+        filter_lookup["created_at__gte"] = last_created_at
+    lines = OrderLine.objects.filter(**filter_lookup)
+    if already_processed_ids:
+        lines = lines.exclude(id__in=already_processed_ids)
+    lines = lines.order_by("created_at")[:ORDER_LINE_DISCOUNT_BATCH_SIZE]
+    discount_line_ids = list(lines.values_list("id", flat=True))
+    if discount_line_ids:
+        order_lines = (
+            OrderLine.objects.filter(id__in=discount_line_ids)
+            .order_by("created_at")
+            .prefetch_related("order")
```

Comment:
Same as comment with `select_related`, cause if I'm not mistaken, we only need `channel_id` to be fetched.

Reply:
We are doing recalculation on `order_lines` so I just leave `select_related` with `order`